neptune-core/src/application/json_rpc/server/service.rs

```diff
@@ -89,17 +99,183 @@ impl RpcApi for RpcServer {
                 .collect(),
         }
     }
+
+    async fn get_block_digest_call(
+        &self,
+        request: GetBlockDigestRequest,
+    ) -> GetBlockDigestResponse {
+        let state = self.state.lock_guard().await;
+
+        GetBlockDigestResponse {
+            digest: request.selector.as_digest(&state).await,
```

Comment:
Danger!!!

You're acquiring the read lock on state twice here. If any other thread requests a write lock after the 1st read lock has been granted but before the 2nd has, that will cause a deadlock. Since no read locks are granted once a write lock is requested.